src/OVAL/probes/independent/yamlfilecontent_probe.c

```diff
@@ -35,11 +37,107 @@
 #include "list.h"
 #include "probe/probe.h"
 
+#define OSCAP_YAML_STRING_TAG "tag:yaml.org,2002:str"
+#define OSCAP_YAML_BOOL_TAG "tag:yaml.org,2002:bool"
+#define OSCAP_YAML_FLOAT_TAG "tag:yaml.org,2002:float"
+#define OSCAP_YAML_INT_TAG "tag:yaml.org,2002:int"
+
+#define OVECCOUNT 30 /* should be a multiple of 3 */
+
 int yamlfilecontent_probe_offline_mode_supported()
 {
 	return PROBE_OFFLINE_OWN;
 }
 
+static bool match_regex(const char *pattern, const char *value)
+{
+	const char *errptr;
+	int erroroffset;
+	pcre *re = pcre_compile(pattern, 0, &errptr, &erroroffset, NULL);
+	if (re == NULL) {
+		dE("pcre_compile failed on pattern '%s': %s at %d", pattern,
+			errptr, erroroffset);
+		return false;
+	}
+	int ovector[OVECCOUNT];
+	int rc = pcre_exec(re, NULL, value, strlen(value), 0, 0, ovector, OVECCOUNT);
+	if (rc > 0) {
+		return true;
+	}
+	return false;
+}
+
+static SEXP_t *yaml_scalar_event_to_sexp(yaml_event_t event)
+{
+	char *tag = (char *) event.data.scalar.tag;
+	char *value = (char *) event.data.scalar.value;
+
+	/* nodes lacking an explicit tag are given a non-specific tag:
+	 * “!” for non-plain scalars, and “?” for all other nodes
+	 */
+	if (tag == NULL) {
+		if (event.data.scalar.style != YAML_PLAIN_SCALAR_STYLE) {
+			tag = "!";
+		} else {
+			tag = "?";
+		}
+	}
+
+	/* Nodes with "!" tag can be sequences, maps or strings, but we process
+	 * only scalars in this functions, so they can only be strings. */
+	if (!strcmp(tag, "!")) {
+		tag = OSCAP_YAML_STRING_TAG;
```

Comment:
Do we expect that the tag can be non-NULL and have tag as `!`?
Otherwise, the assignment `tag = OSCAP_YAML_STRING_TAG;` can be done in line 80.

Reply:
Yes, according to https://yaml.org/spec/1.2/ section 6.9.1:

It is possible for the tag property to be explicitly set to the “!” non-specific tag. By convention, this “disables” tag resolution, forcing the node to be interpreted as “tag:yaml.org,2002:seq”, “tag:yaml.org,2002:map”, or “tag:yaml.org,2002:str”, according to its kind. 

There is no way to explicitly specify the “?” non-specific tag. This is intentional.

Reply:
Thanks for the clarification.